rfcs/20240315-algorithm-support.md

```diff
@@ -0,0 +1,166 @@
+| (I8)  | `algorithm` | an enum of `UNSET`, `ANY_F8_ANY_F8_F32`, `ANY_F8_ANY_F8_F32_FAST_ACCUM`, `F16_F16_F16`, `F16_F16_F32`, `BF16_BF16_BF16`, `BF16_BF16_F32`, `BF16_BF16_F32_X3`, `BF16_BF16_F32_X6`, `TF32_TF32_F32`, `TF32_TF32_F32_X3`, `F32_F32_F32`, and `F64_F64_F64` |             |
+
+#### Examples
+
+*Change the example:*
+
+```mlir
+// %lhs: [
+//        [[1, 2],
+//         [3, 4]],
+//        [[5, 6],
+//         [7, 8]]
+//       ]
+// %rhs: [
+//        [[1, 0],
+//         [0, 1]],
+//        [[1, 0],
+//         [0, 1]]
+//       ]
+%result = "stablehlo.dot_general"(%lhs, %rhs) {
+  dot_dimension_numbers = #stablehlo.dot<
+    lhs_batching_dimensions = [0],
+    rhs_batching_dimensions = [0],
+    lhs_contracting_dimensions = [2],
+    rhs_contracting_dimensions = [1]
+  >,
+  precision_config = [#stablehlo<precision DEFAULT>, #stablehlo<precision DEFAULT>],
+  algorithm = #stablehlo<dot_algorithm UNSET>
+} : (tensor<2x2x2xi64>, tensor<2x2x2xi64>) -> tensor<2x2x2xi64>
+// %result: [
+//           [[1, 2],
+//            [3, 4]],
+//           [[5, 6],
+//            [7, 8]]
+//          ]
+```
+
+## Alternatives considered
+
+### Adding new Precision values instead of the algorithm property
+
+The precision config is per operand, but the algorithms describe multiple
+properties of the computation, some of which are not connected to the operands
+(such as the accumulation type, number of dot operations to use). So we think
+that it's more conceptually correct to add a separate algorithm property.
+
+### Making the algorithm property a struct instead of an enum
+
+The algorithm property describes multiple properties of the computation: AType,
+BType, AccumType, NumOps, and occasionally other options such as FastAccum. It
+maybe easier to process if we had a "struct" to describe the algorithm, instead
+of a name. For example:
+`{AType=BF16, BType=BF16, AccumType=F32, NumOps=6, FastAccum=False}`
+instead of `BF16_BF16_F32_X6`. This could also be
+useful if we introduced algorithms which have asymmetric input precisions (but
+the same input storage types), to disambiguate which operand has which
+precision.
+
+Having considered the mentioned advantages, we wanted to start by adding
+something more simple for now, and just add support for a few algorithms which
+are important to disambiguate. (All earlier types and precisions would remain
+supported for now). Also, we think that algorithm names are easier to refer to
+(although this could be probably mitigated by introducing a special syntax to
+write the structs). We also think that algorithms may have more different
+properties which would need new struct members, but if we use an enum, we can
+just append them to the name.
```

Comment:
+1, also structs may require more verification if not all combinations are valid.

Reply:
We held a meeting on the 9th of May, and had a discussion with @majnemer, where he advocated for using a struct instead, and he was quite convincing.
I think his main points were that:
- This could prevent combinatorial explosion of the enum types.
- The current enum based solution could get messy if we'll have examples where the "precision" of the lhs and rhs are not the same.

I guess that I'll try to write up how would that look like.
I think that it's OK that HLO uses enums now, we could handle this in the converter for now if we'll use a struct in StableHLO.

Reply:
I was thinking of this:

```
struct DotAlgorithm {
  mlir::Type lhs_type; // Meaning that the LHS will be "casted" to this type before doing the operation. We could also call this a_type?
  mlir::Type rhs_type;
  mlir::Type accum_type;
  int lhs_component_count; // This can be >1 if we decompose the LHS to do an approximation using multiple primitive operations.
  int rhs_component_count;
  int num_primitive_ops; // This can be >1 if we decompose the LHS to do an approximation using multiple primitive operations.
  bool can_accumulate_in_lower_precision_for_some_steps; // We can refine this later. Corresponds to CUBLASLT_MATMUL_DESC_FAST_ACCUM.
};
```

Not sure if it should be DotAlgorithm or just Algorithm, not sure how well these generalize.

Here is how the previously proposed enum values could be described as structs:

```
TF32_TF32_F32 -> 
{lhs_type: TF32, 
 rhs_type: TF32,
 accum_type: F32,
 lhs_component_count: 1,
 rhs_component_count: 1,
 num_primitive_ops: 1,
 can_accumulate_in_lower_precision_for_some_steps: false}


BF16_BF16_F32_X6 ->
{lhs_type: BF16,
 rhs_type: BF16,
 accum_type: F32,
 lhs_component_count: 3,
 rhs_component_count: 3,
 num_primitive_ops: 6,
 can_accumulate_in_lower_precision_for_some_steps: false}


ANY_F8_ANY_F8_F32_FAST_ACCUM -> For example:
{lhs_type: F8E5M2,
 rhs_type: F8E5M2,
 accum_type: F32,
 lhs_component_count: 3,
 rhs_component_count: 3,
 num_primitive_ops: 6,
 can_accumulate_in_lower_precision_for_some_steps: true}
```

Regarding the verification:
Maybe we can do it in the stable_hlo to hlo conversion?

Reply:
>> Regarding the verification:

There does not seem to be any constraint among the different fields of the struct. If there are then prefer does the verification checks in StableHLO.

Reply:
Updated the RFC to reflect these changes.